File: leecode/code/98/98.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {

    public boolean vall(TreeNode root, Integer lower, Integer higher) {
        if (root == null)
            return true;

        int temp = root.val;

        if (lower != null && temp <= lower) return false;
        if (higher != null && temp >= higher) return false;

        if (!vall(root.left, lower, temp)) return false;
        if (!vall(root.right, temp, higher)) return false;
        return true;

    }

    public boolean isValidBST(TreeNode root) {
        return vall(root, null, null);

    }
}

File: leecode/code/98/98.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:


    def isValidBST(self, root: TreeNode) -> bool:
        
        def helper(root, lower = float("inf"), upper = float("inf")):
            if root == None:
                return True
            if lower != float("inf") and root.val <= lower :
                return False
            if upper != float("inf") and root.val >= upper:
                return False


            if not helper(root.left, lower, float(root.val)):
                return False
            if not helper(root.right, float(root.val), upper):
                return False
            return True


        return helper( root)